### PR TITLE
[SYCL-MLIR] Hack to allow polygeist to reuse the clang backend utils

### PR DESCRIPTION
**Pull request by Naghasan**
Hack to allow cgiest to use the clang routine to setup the llvm pipeline rather than its own. This allows to capture the SYCL specific pass setup and allows cgeist to honor clang's codegen options.

The patch also fixes a small issue in handling the libsycl unbundling.

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -9249,8 +9255,16 @@ void OffloadBundler::ConstructJobMultipleOutputs(
...
+    } else {
+      if (Dep.DependentToolChain->getTriple().getEnvironment() ==
+          llvm::Triple::SYCLMLIR) {
+        llvm::Triple Tmp{Dep.DependentToolChain->getTriple().getArchName(),
+                         Dep.DependentToolChain->getTriple().getVendorName(),
+                         Dep.DependentToolChain->getTriple().getOSName()};
+        Triples += Tmp.normalize();
+      } else
+        Triples += Dep.DependentToolChain->getTriple().normalize();
+    }
```
Comment (victor-eds) [inline clang/lib/Driver/ToolChains/Clang.cpp:9267]:
```suggestion
    } else if (Dep.DependentToolChain->getTriple().getEnvironment() ==
                   llvm::Triple::SYCLMLIR) {
        llvm::Triple Tmp{Dep.DependentToolChain->getTriple().getArchName(),
                         Dep.DependentToolChain->getTriple().getVendorName(),
                         Dep.DependentToolChain->getTriple().getOSName()};
        Triples += Tmp.normalize();
    } else
      Triples += Dep.DependentToolChain->getTriple().normalize();
```

**Review by victor-eds — COMMENTED** [review]
Just a minor

**Review by whitneywhtsang — APPROVED** [review]

**Review by etiotto — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
Should I do a SYCL-Bench run after this PR get merged? For the paper, are there any other PRs we have pending that have performance impact on SYCL-Bench other than this one?

**Review by victor-eds — APPROVED** [review]

**Comment by Naghasan** [comment]
@whitneywhtsang I don't think you'll get drastically different figures with just that patch (the annoying code patterns are still there). This patch will however support the constant propagation Victor P is doing.

File: polygeist/tools/cgeist/driver.cc [hunk 2]
```diff
@@ -900,64 +931,70 @@ static LogicalResult compileModule(mlir::OwningOpRef<mlir::ModuleOp> &Module,
...
 
     LLVMModule->setDataLayout(DL);
     LLVMModule->setTargetTriple(Triple.getTriple());
+
     LLVM_DEBUG(llvm::dbgs()
                << "*** Translated MLIR to LLVM IR successfully ***\n");
 
-    if (EmitBC || EmitLLVM) {
-      // Not needed when emitting binary for now; will be handled by the driver.
-      runOptimizationPipeline(*LLVMModule, options);
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/driver.cc:913]: After this change:
```
/iusers/waihungt/llvm/polygeist/tools/cgeist/driver.cc:886:13: warning: ‘void runOptimizationPipeline(llvm::Module&, Options&)’ defined but not used [-Wunused-function]
  886 | static void runOptimizationPipeline(llvm::Module &Module, Options &options) {
      |             ^~~~~~~~~~~~~~~~~~~~~~~
```
Is `runOptimizationPipeline` no longer needed?

Comment (Naghasan) [reply polygeist/tools/cgeist/driver.cc:913]: yes... I'll make a PR to remove that

Comment (Naghasan) [reply polygeist/tools/cgeist/driver.cc:913]: https://github.com/intel/llvm/pull/10686

**Pull request merged**